New Vite plugin docs for local development (#20586)
cloudflare/cloudflare-docs: src/content/docs/workers/local-development/index.mdx

---
pcx_content_type: navigation
title: Local development
sidebar:
  order: 5
head: []
description: Develop your Workers locally.
---

Cloudflare Workers provides two primary ways to develop and test your code locally. Local development ensures a fast feedback loop, keeps iteration times low, and lets you experiment before deploying to production.

> Developers can locally test their changes to a Workers project using the following:

Reviewer: ```suggestion
You can test changes locally using the following:
```

> Developers can locally test their changes to a Workers project using the following:

Reviewer: Some other details on this page that could be worked in:

- ref what frameworks integrate with the vite plugin
- some guidance on when you should use which
- any compat matrix style differences (at the most basic, that the vite plugin only is for Workers)

> - **Wrangler’s built-in dev command** [`wrangler dev`](/workers/wrangler/commands/#dev), which uses [`workerd`](https://github.com/cloudflare/workerd) and Miniflare to simulate production behavior and resource bindings locally.

Reviewer: Seems helpful to the reader to find some way to call out remote dev? Since only possible with Wrangler?

> - The **Cloudflare Vite plugin** (`@cloudflare/vite-plugin`), which integrates directly with Vite and runs your code in the same `workerd` runtime, while also offering features like [hot module replacement (HMR)](https://vite.dev/guide/features.html#hot-module-replacement). The Vite plugin is currently considered in **beta**.

Reviewer: ```suggestion
- The **Cloudflare Vite plugin** (`@cloudflare/vite-plugin`), which integrates directly with Vite and runs your code in the [Worker Runtime](https://developers.cloudflare.com/workers/reference/how-workers-works/) ([`workerd`](https://github.com/cloudflare/workerd)), while also offering features like [hot module replacement (HMR)](https://vite.dev/guide/features.html#hot-module-replacement). The Vite plugin is currently in open beta.
```

---
pcx_content_type: navigation
title: Local development
sidebar:
  order: 5
head: []
description: Develop your Workers locally.
---

> Cloudflare Workers provides two primary ways to develop and test your code locally. Local development ensures a fast feedback loop, keeps iteration times low, and lets you experiment before deploying to production.
> 
> You can test changes locally using the following:

Reviewer 2: ```suggestion
There are two primary ways to develop Cloudflare Workers locally. Local development ensures a fast feedback loop, keeps iteration times low, and lets you experiment before deploying to production.

You can develop locally using the following:
```

> - **Wrangler’s built-in dev command** [`wrangler dev`](/workers/wrangler/commands/#dev), which runs your code in the [Workers Runtime](/workers/reference/how-workers-works/) ([`workerd`](https://github.com/cloudflare/workerd)) to simulate production behavior and resource bindings locally. Wrangler also offers the option to [develop against remote resources](/workers/local-development/wrangler/#develop-using-remote-resources-and-bindings) with `wrangler dev --remote`, allowing you to test directly with data stored on Cloudflare's network.

Reviewer 2: ```suggestion
- **Wrangler’s built-in dev command** [`wrangler dev`](/workers/wrangler/commands/#dev), which runs your code in the [Workers Runtime](/workers/reference/how-workers-works/) ([`workerd`](https://github.com/cloudflare/workerd)) to simulate production behavior and resource bindings locally. Wrangler also offers the option to [develop against remote resources](/workers/local-development/wrangler/#develop-using-remote-resources-and-bindings) with `wrangler dev --remote`, allowing you to work directly with data stored on Cloudflare's network.
```

Generally trying to avoid using the word 'test', which could be misinterpreted.

- The **Cloudflare Vite plugin** (`@cloudflare/vite-plugin`), which integrates directly with [Vite](https://vite.dev/) and runs your code in the [Workers Runtime](/workers/reference/how-workers-works/) ([`workerd`](https://github.com/cloudflare/workerd)) to simulate production behavior and resource bindings locally. The Vite plugin also offers features like [hot module replacement (HMR)](https://vite.dev/guide/features.html#hot-module-replacement) and access to Vite's ecosystem and build time transformations. It is currently in **open beta**.

> Both approaches aim to provide confidence that your local environment will closely match the production runtime, removing the need to [test against remote resources](#develop-using-remote-resources-and-bindings).

Reviewer 2: ```suggestion
Both approaches enable accessing local resource bindings and aim to provide confidence that your local environment closely matches production.
```

## Choosing Between Wrangler and Vite

Deciding between Wrangler and the Cloudflare Vite plugin depends on your project's focus and development workflow. Here are some quick guidelines to help you choose:

### When to use Wrangler:

- **Backend & Workers-focused:**
  If you're primarily building APIs, serverless functions, or background tasks that need to closely match production behavior, use Wrangler. Wrangler offers the most accurate local simulation of Cloudflare’s runtime.

- **Real Cloudflare resources:**
  Need to test actual Cloudflare data or services? Wrangler’s `--remote` flag lets you do so without fully deploying to production.

- **Simple frontends:**

>   If you have minimal frontend requirements and don’t need hot reloading or advanced bundling, Wrangler by itself is sufficient.

Reviewer 2: ```suggestion
  If you have minimal frontend requirements and don’t need hot reloading or advanced bundling, Wrangler may be sufficient.
```

### When to use the Cloudflare Vite Plugin:

- **Frontend-centric development:**

>   If you rely on modern frameworks like React, Vue, Svelte, or Solid, the Vite plugin simplifies setup and integration.

Reviewer 2: ```suggestion
  If you already use Vite with modern frontend frameworks like React, Vue, Svelte, or Solid, the Vite plugin integrates into your development workflow.
```

> 

Reviewer 2: ```suggestion

- **React Router v7**
  If you are using [React Router v7](https://reactrouter.com/) (the successor to Remix), it is officially supported by the Vite plugin as a full-stack SSR framework.
  
```

---
pcx_content_type: navigation
title: Local development
sidebar:
  order: 6
head: []
description: Develop your Workers locally.
---

import { Details, LinkCard, Render, PackageManagers } from "~/components";

When building projects on Cloudflare Workers, you have two options for local development:

- [**Wrangler**](/workers/wrangler/), using the built-in [`wrangler dev`](/workers/wrangler/commands/#dev) command.
- [Vite](https://vite.dev/), using the [**Cloudflare Vite plugin**](/workers/vite-plugin/).

> Both Wrangler and the Vite plugin use [Miniflare](/workers/testing/miniflare/) to provide an accurate **local** simulation of the Cloudflare Workers runtime, ([`workerd`](https://github.com/cloudflare/workerd)). If you need to [develop with **remote resources**](/workers/local-development/remote-data/), Wrangler is the only option that provides remote development with the `wrangler dev --remote` command.

Reviewer 2: ```suggestion
Both Wrangler and the Vite plugin use [Miniflare](/workers/testing/miniflare/) to provide an accurate **local** simulation of the Cloudflare Workers runtime, ([`workerd`](https://github.com/cloudflare/workerd)). If you need to [develop with **remote resources**](/workers/local-development/remote-data/), Wrangler is the only option, and provides remote development via the `wrangler dev --remote` command.
```

<Render file="install_wrangler" />

#### Start a development server

To start a development server using Wrangler, run the following:

<PackageManagers type="exec" pkg="wrangler" args="dev" />

<LinkCard title="Learn more about Wrangler" href="/workers/wrangler/" />

### Vite Plugin

#### Installation

To install the Cloudflare Vite plugin, first start with a simple `package.json`

```json title="package.json"
{
	"name": "cloudflare-vite-get-started",
	"private": true,
	"version": "0.0.0",
	"type": "module",
	"scripts": {
		"dev": "vite dev",
		"build": "vite build",
		"preview": "npm run build && vite preview",
		"deploy": "npm run build && wrangler deploy"
	}
}
```

Then, install the necessary development depenencies:

<PackageManagers type="add" pkg="vite @cloudflare/vite-plugin wrangler" dev />

This command will install `vite`, along with `@cloudflare/vite-plugin` and `wrangler` as `devDependencies` in your `package.json`.

#### Start a development server

You can then start a development server by running:

<PackageManagers type="run" args="dev" />

<LinkCard
	title="Learn more about the Vite plugin"
	href="/workers/vite-plugin/"

> />

Reviewer 2: I'm not sure it makes sense to include these setup examples as both of them will fail to run. At minimum, you would need to add steps to include a `wrangler.jsonc` config file and Worker entry file, but then you're duplicating the documentation that is elsewhere.

My preference would be to remove this section and replace it with the content from the 'Wrangler vs Vite' page. Then there would be no need to have that as a separate page and the `LinkCard` above can also be removed.
